ide/html.editor/src/org/netbeans/modules/html/editor/hints/other/AddMissingAltAttributeHint.java

```diff
@@ -0,0 +1,107 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.html.editor.hints.other;
+
+import java.awt.EventQueue;
+import java.util.Collections;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.swing.text.BadLocationException;
+import javax.swing.text.Document;
+import org.netbeans.editor.BaseDocument;
+import org.netbeans.modules.csl.api.Hint;
+import org.netbeans.modules.csl.api.HintFix;
+import org.netbeans.modules.csl.api.OffsetRange;
+import org.netbeans.modules.html.editor.hints.HtmlRuleContext;
+import org.netbeans.modules.html.editor.utils.HtmlTagContextUtils;
+import org.netbeans.modules.parsing.api.Source;
+
+/**
+ *
+ * @author Christian Lenz
+ */
+public class AddMissingAltAttributeHint extends Hint {
+
+    public AddMissingAltAttributeHint(HtmlRuleContext context, OffsetRange range) {
+        super(AddMissingAltAttributeRule.getInstance(),
+            AddMissingAltAttributeRule.getInstance().getDescription(),
+            context.getFile(),
+            range,
+            Collections.<HintFix>singletonList(new AddMissingAltAttributeHintFix(context, range)),
+            10);
+    }
+
+    private static class AddMissingAltAttributeHintFix implements HintFix {
+
+        private static final Logger LOGGER = Logger.getLogger(AddMissingAltAttributeHintFix.class.getSimpleName());
+
+        HtmlRuleContext context;
+        OffsetRange range;
+
+        public AddMissingAltAttributeHintFix(HtmlRuleContext context, OffsetRange range) {
+            this.context = context;
+            this.range = range;
+        }
+
+        @Override
+        public String getDescription() {
+            return AddMissingAltAttributeRule.getInstance().getDisplayName();
+        }
+
+        @Override
+        public void implement() throws Exception {
+            BaseDocument document = (BaseDocument) context.getSnapshot().getSource().getDocument(true);
+            document.runAtomic(() -> {
+                try {
+                    OffsetRange adjustedRange = HtmlTagContextUtils.adjustContextRange(document, range.getStart(), range.getEnd(), true);
+                    String tagContent = document.getText(adjustedRange.getStart(), adjustedRange.getLength());
+
+                    // Find last self-closing or non self-closing tag
+                    Pattern closingTagPattern = Pattern.compile("(/?>)");
+                    Matcher closingTagMatcher = closingTagPattern.matcher(tagContent);
+
+                    if (closingTagMatcher.find()) {
+                        int altInsertPosition = adjustedRange.getStart() + closingTagMatcher.start(1);
```

Comment:
This might lead to wrong results. Consider PHP with embedded HTML. The document could look like this:

```php
<img src='demo.png' alt='<?php echo htmlspecialchars('x > y');?>' />
```

I would use the HTML token stream and either use the start of the closing token or, if that does not exist, the start of the last HTML token + length.

Reply:
TBH, I didn't try PHP yet, will do it too. Thx for the hint. If I guess it correct, your example will not trigger the hint, because you already have the alt attribute. But yes It would be better to take the HTML token instead of just looking for the closing tag symbol with an regex.